js/viz/funnel/funnel.js

```diff
@@ -0,0 +1,245 @@
+    },
+
+    _eventsMap: {
+        onHoverChanged: { name: "hoverChanged" },
+        onSelectionChanged: { name: "selectionChanged" }
+    },
+
+    _disposeCore: function() {
+    },
+
+    _applySize: function(rect) {
+        this._rect = rect.slice();
+        this._change(["TILING"]);
+    },
+
+    _change_TILING: function() {
+        var that = this,
+            items = that._items,
+            rect = that._rect;
+
+        this._group.clear();
+
+        items.forEach(function(item, index) {
+            var coords = item.figure.map(function(c, index) {
+                    var offset = index % 2;
+                    return rect[0 + offset] + (rect[2 + offset] - rect[0 + offset]) * c;
+                }),
+                element = that._renderer.path([], "area")
+                    .attr({
+                        points: coords
+                    })
+                    .append(that._group);
+
+            item.coords = coords;
+            item.element = element;
+        });
+
+        this._requestChange(["TILES"]);
+    },
+
+    _customChangesOrder: ["NODES_CREATE", "LAYOUT", "TILING", "TILES"],
+
+    _dataSourceChangedHandler: function() {
+        this._requestChange(["NODES_CREATE"]);
+    },
+
+    _change_DATA_SOURCE: function() {
+        this._updateDataSource();
+    },
+
+    _change_NODES_CREATE: function() {
+        this._buildNodes();
+    },
+
+    _change_TILES: function() {
+        this._applyTilesAppearance();
+    },
+
+    _suspend: function() {
+        if(!this._applyingChanges) {
+            this._suspendChanges();
+        }
+    },
+    _resume: function() {
+        if(!this._applyingChanges) {
+            this._resumeChanges();
+        }
+    },
+
+    _applyTilesAppearance: function() {
+        this._items.forEach(function(item) {
+            var state = item.getState();
+            item.element.smartAttr(item.states[state]);
+        });
+    },
+
+    _getProxyData: function(x, y) {
+        var that = this,
+            data;
+
+        this._proxyData.some(function(callback) {
+            data = callback.call(that, x, y);
+            if(data) {
+                return true;
+            }
+        });
+
+        return data;
+    },
+
+    clearHover: function() {
+        this._suspend();
+        this._items.forEach(function(item) {
+            item.isHovered() && item.hover(false);
+        });
+        this._resume();
+    },
+
+    clearSelection: function() {
+        this._suspend();
+        this._items.forEach(function(item) {
+            item.isSelected() && item.select(false);
+        });
+        this._resume();
+    },
+
+    _getData: function() {
+        var that = this,
+            data = that._dataSourceItems() || [],
+            valueField = that._getOption("valueField", true),
+            argumentField = that._getOption("argumentField", true),
+            colorField = that._getOption("colorField", true),
+            items = data.map(function(item) {
```

Comment:
I know that data might not be very large, but use single reduce instead of pair of map and filter